services/qan/client.go

```diff
@@ -77,32 +78,33 @@ func mergeLabels(node *models.Node, service *models.Service, agent *models.Agent
 }
 
 // Collect adds labels to the data from pmm-agent and sends it to qan-api.
-func (c *Client) Collect(ctx context.Context, req *qanpb.CollectRequest) error {
+func (c *Client) Collect(ctx context.Context, metricsBuckets []*agentpb.MetricsBucket) error {
 	// TODO That code is simple, but performance will be very bad for any non-trivial load.
 	// https://jira.percona.com/browse/PMM-3894
 
 	start := time.Now()
 	defer func() {
 		if dur := time.Since(start); dur > time.Second {
-			c.l.Warnf("Collect for %d buckets took %s.", len(req.MetricsBucket), dur)
+			c.l.Warnf("Collect for %d buckets took %s.", len(metricsBuckets), dur)
 		}
 	}()
+	metricBuckets := make([]*qanpb.MetricsBucket, len(metricsBuckets))
```

Comment:
First, it is a bad name: `metricBuckets` and `metricsBuckets` are too close to each other.

Second, there is a logical mistake: we make `metricBuckets` of the same length as `metricsBuckets`, but then we can skip (`continue`) some buckets below. That will result in nil values at the end of the slice.

I think we should set capacity, not length there, and `append` to the slice.

Please also add a test for that.

Reply:
> Please also add a test for that.

https://github.com/percona/pmm-managed/pull/230/commits/cfd30e24ce0810a1bfc971a965cf793c4105e393

Reply:
sorry, missed that